openpgp/ocb/ocb.go

```diff
@@ -0,0 +1,279 @@
+// Copyright (C) 2019 ProtonTech AG
+
+// Package ocb provides an implementation of the OCB (offset codebook) mode of
+// operation, as described in RFC-7253 of the IRTF and in Rogaway, Bellare,
+// Black and Krovetz - OCB: A BLOCK-CIPHER MODE OF OPERATION FOR EFFICIENT
+// AUTHENTICATED ENCRYPTION (2003).
+// Security considerations (from RFC-7253): A private key MUST NOT be used to
+// encrypt more than 2^48 blocks. Tag length should be at least 12 bytes (a
+// brute-force forging adversary succeeds after 2^{tag length} attempts). A
+// single key SHOULD NOT be used to decrypt ciphertext with different tag
+// lengths. Nonces need not be secret, but MUST NOT be reused.
+// This package only supports underlying block ciphers with 128-bit blocks,
+// such as AES-{128, 192, 256}, but may be extended to other sizes.
+
+package ocb
+
+import (
+	"math/bits"
+	"crypto/cipher"
+	"golang.org/x/crypto/openpgp/bytesutils"
+	"bytes"
+	"errors"
+)
+
+type ocb struct {
+	block     cipher.Block
+	tagSize   int
+	nonceSize int
+	mask      mask
+}
+
+type mask struct {
+	// L_*, L_$, (L_i)_{i ∈ N}
+	L_ast []byte
+	L_dol []byte
+	L     [][]byte
+}
+
+const (
+	defaultTagSize     = 16
+	defaultNonceSize   = 15
+)
+
+// NewOCB returns an OCB instance with the given block cipher and default
+// tag and nonce sizes.
+func NewOCB(block cipher.Block) ocb {
+
+	return ocb{
+		block:     block,
+		tagSize:   defaultTagSize,
+		nonceSize: defaultNonceSize,
+		mask:      initializeMaskTable(block),
+	}
```

Comment:
Perhaps change this to
```go
NewOCBWithNonceAndTagSize(block, defaultNonceSize, defaultTagSize)
```
Also because, as written, you're missing the block size check :)

Reply:
Good idea and nicely spotted